app/handlers/contact_infos_confirm.rb

```diff
@@ -15,6 +15,10 @@ def handle
     fatal_error(code: :no_contact_info_for_code) if params[:code].nil?
     contact_info = ContactInfo.where(confirmation_code: params[:code]).first
     fatal_error(code: :no_contact_info_for_code) if contact_info.nil?
+    user = contact_info.user
+    # the came_from url is only used when user is signing up
+    outputs[:stored_url] = user.came_from if user.is_temp?
```

Comment:
I would prefer for this `user.is_temp?` logic to not be here as it isn't the primary responsibility of this routine (which should pretty much just be confirming emails).  Since you've added `contact_info` as an output, seems like the controller can do this work.

And if it is true that `came_from` is only set during sign up, can we rename it to reflect that?  E.g. `registration_redirect_url`.  Then if that is not nil, we know we are in the middle of the registration process and can redirect accordingly.

```ruby
class User 
  def registration_redirect_url
    is_temp? ? nil : read_attribute(:registration_redirect_url)
  end
  ...
```

```ruby
handle_with(ContactInfosConfirm,
            complete: lambda {
              # If in the middle of registration, redirect to continue process
              redirect_url = @handler_result.outputs.contact_info.user.registration_redirect_url
              if redirect_url.present?
                redirect_to redirect_url,
                            notice: 'Thanks for adding your email address.'
              else
                render :confirm, status: @handler_result.errors.any? ? 400 : 200
              end
            })
```